### PR TITLE
refactor: convert RepoType from int to interface

### PR DESCRIPTION
**Pull request by nonsense**
## Related Issues
This PR is convering `RepoType` type from `int` to `interface`, so that we can use the `repo` package in other projects and provide a custom Config, that is not referred to in Lotus (such as `config.Boost`).

The new usage can be seen at: https://github.com/filecoin-project/boost/pull/231

---

We want to be using the `filecoin-project/lotus/node/repo` package, due to the fx dependency injection, as we want to integrate all Lotus markets dependencies directly in boost before releasing the alpha version.

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [x] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] CI is green

### CODE REVIEW

**Review by dirkmc — APPROVED** [review]

File: node/repo/memrepo.go [hunk 1]
```diff
@@ -106,7 +106,7 @@ func (lmem *lockedMemRepo) Path() string {
 		panic(err) // only used in tests, probably fine
 	}
 
-	if lmem.t == StorageMiner {
+	if lmem.t.Type() == "StorageMiner" || lmem.t.Type() == "Boost" {
```
Comment (magik6k) [inline node/repo/memrepo.go:109]: I'd use trait-interfaces instead of hard-coding (so this would be something like `if lmem.t.(interface{HasStorage()})`, and maybe a separate trait for `deal-staging`)

File: node/repo/fsrepo.go [hunk 2]
```diff
@@ -41,47 +41,83 @@ const (
...
-		"StorageMiner",
-		"Worker",
-		"Wallet",
-		"Markets",
-	}
-	if t < 0 || int(t) > len(s) {
-		return "__invalid__"
-	}
-	return s[t]
+type FullNodeRepoType struct {
```
Comment (magik6k) [inline node/repo/fsrepo.go:64]: The `RepoType` suffix feels a bit redundant.

Comment (magik6k) [reply node/repo/fsrepo.go:64]: Also given that all of those structs are field-less it could be nicer to do

```
type fullNode struct {}
var FullNode fullNode
```

This way we don't need to add `{}` everywhere

File: node/repo/fsrepo.go [hunk 3]
```diff
@@ -41,47 +41,83 @@ const (
...
+
+func (f WalletRepoType) Type() string {
+	return "Wallet"
+}
+
+func (f WalletRepoType) Config() interface{} {
+	return &struct{}{}
 }
 
 func defConfForType(t RepoType) interface{} {
```
Comment (magik6k) [inline node/repo/fsrepo.go:119]: Can drop this function now

File: node/modules/storageminer_dagstore.go [hunk 4]
```diff
@@ -23,88 +23,70 @@ const (
...
-		return nil, err
-	}
-
-	// caps the amount of concurrent calls to the storage, so that we don't
-	// spam it during heavy processes like bulk migration.
-	if v, ok := os.LookupEnv("LOTUS_DAGSTORE_MOUNT_CONCURRENCY"); ok {
-		concurrency, err := strconv.Atoi(v)
-		if err == nil {
-			cfg.MaxConcurrencyStorageCalls = concurrency
+func NewMinerAPI(cfg config.DAGStoreConfig) func(fx.Lifecycle, repo.LockedRepo, dtypes.ProviderPieceStore, mdagstore.SectorAccessor) (mdagstore.MinerAPI, error) {
```
Comment (magik6k) [inline node/modules/storageminer_dagstore.go:26]: Unrelated?

Comment (nonsense) [reply node/modules/storageminer_dagstore.go:26]: This gets rid of the `extractDAGStoreConfig` which casts the `config`, so it is somewhat related and necessary as part this work, but I agree slightly tangential to the rest of the changes.

File: node/repo/memrepo.go (hunk 1 above)
Comment (nonsense) [reply node/repo/memrepo.go]: Check last commit, and let me know if this is what you had in mind: https://github.com/filecoin-project/lotus/pull/8086/commits/9fd7e71a07c0772d1a196c4a53fa72140dacf189

**Comment by nonsense** [comment]
Thanks for review @magik6k !!! I think I addressed all your comments.

**Comment by nonsense** [comment]
@magik6k we modified a bit further the config in order to be able to pass in `config.Boost`. I am not happy with how we handle the dynamic configs... any suggestions on how to do that in a better way?

File: node/config/types.go [hunk 5]
```diff
@@ -52,6 +52,32 @@ type StorageMiner struct {
...
+}
+
+func (c *StorageMiner) GetSealingConfig() SealingConfig {
+	return c.Sealing
+}
+
+func (c *StorageMiner) SetSealingConfig(other SealingConfig) {
+	c.Sealing = other
+}
+
```
Comment (magik6k) [inline node/config/types.go:80]: This file should only contain config structs with docstrings, putting other stuff here will mess with config docsgen

File: node/repo/fsrepo.go [hunk 6]
```diff
@@ -41,47 +41,94 @@ const (
...
+	return "StorageMiner"
+}
+
+func (storageMiner) Config() interface{} {
+	return config.DefaultStorageMiner()
+}
+
+var Markets markets
+
+type markets struct{}
```
Comment (magik6k) [inline node/repo/fsrepo.go:98]: Shouldn't this also have `SupportsStagingDeals`?

File: node/builder.go [hunk 7]
```diff
@@ -224,12 +224,16 @@ var LibP2P = Options(
 )
 
 func IsType(t repo.RepoType) func(s *Settings) bool {
-	return func(s *Settings) bool { return s.nodeType == t }
+	return func(s *Settings) bool { return s.nodeType.Type() == t.Type() }
```
Comment (magik6k) [inline node/builder.go:227]: Do we actually need the `.Type()` here?

File: cli/util/api.go [hunk 8]
```diff
@@ -32,14 +32,14 @@ const (
 // of the API server (only used by the tests), in the order of precedence they
 // should be applied for the requested kind of node.
 func flagsForAPI(t repo.RepoType) []string {
```
Comment (magik6k) [inline cli/util/api.go:34]: Could we change those repoType->thing functions into methods on RepoType? This way all of those per-type things are in one place.

File: node/config/types.go (hunk 5 above)
Comment (magik6k) [reply node/config/types.go:80]: (see the comment on top of the file)

File: node/builder.go (hunk 7 above)
Comment (nonsense) [reply node/builder.go]: Now that we have singleton for `nodeType`, we don't. Removing it.

**Comment by nonsense** [comment]
@magik6k I think I addressed the latest comments.

**Comment by nonsense** [comment]
@magik6k I rebased to master, and I think this is ready for merge. Could you have another look, and let me know if you think we need anything else as part of this PR?

**Review by magik6k — APPROVED** [review]

**Comment by nonsense** [comment]
thanks!! :)
